### PR TITLE
Add disk space check before downloading GCS file in GCSHook

### PR DESCRIPTION
**Pull request by MonsterChenzhuo**
---

This PR implements a disk space check in `GCSHook.download` method before initiating the download, addressing a TODO in the code. This prevents runtime errors when downloading large files to a volume with insufficient space.

##### Was generative AI tooling used to co-author this PR?

- [ ] Yes (please specify the tool below)

---

* Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information. Note: commit author/co-author name and email in commits become permanently public when merged.
* For fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
* When adding dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
* For significant user-facing changes create newsfragment: `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

### CODE REVIEW

File: providers/google/src/airflow/providers/google/cloud/hooks/gcs.py [hunk 1]
```diff
@@ -347,6 +344,16 @@ def download(
                 blob = bucket.blob(blob_name=object_name, chunk_size=chunk_size)
 
                 if filename:
+                    blob.reload(timeout=timeout)
+                    if blob.size:
```
Comment (SameerMesiah97) [inline providers/google/src/airflow/providers/google/cloud/hooks/gcs.py]: What about situations where blob.size is 0? Since 0 is falsy, won't this result in the disk size check being skipped? I understand that this will not have a functional impact, but we can be more explicit here. I think this would be more precise:

`if blob.size is not None:`

File: providers/google/tests/unit/google/cloud/hooks/test_gcs.py [hunk 2]
```diff
@@ -1843,3 +1846,88 @@ def get_logs_string(call_args_list):
...
+
+        # Free space 500 bytes
+        # shutil.disk_usage returns a named tuple (total, used, free)
+        mock_disk_usage.return_value = mock.Mock(total=2000, used=1500, free=500)
+
+        with pytest.raises(AirflowException, match="Insufficient disk space"):
+            hook.download(bucket_name="bucket", object_name="object", filename="/tmp/file")
+
+        # Verify blob.reload() was called to get the size
+        mock_blob.reload.assert_called_once()
```
Comment (SameerMesiah97) [inline providers/google/tests/unit/google/cloud/hooks/test_gcs.py:1877]: I would add assertions for `mock_disk_usage` and `download_to_filename` as well. Like this:

`mock_disk_usage.assert_called_once()`
`mock_blob.download_to_filename.assert_not_called()`

File: providers/google/tests/unit/google/cloud/hooks/test_gcs.py [hunk 3]
```diff
@@ -1843,3 +1846,88 @@ def get_logs_string(call_args_list):
...
+        # Free space 2000 bytes
+        mock_disk_usage.return_value = mock.Mock(total=4000, used=2000, free=2000)
+
+        filename = "/tmp/file"
+        hook.download(bucket_name="bucket", object_name="object", filename=filename)
+
+        # Verify blob.reload() was called
+        mock_blob.reload.assert_called_once()
+        # Verify download was called
+        mock_blob.download_to_filename.assert_called_once_with(filename, timeout=mock.ANY)
```
Comment (SameerMesiah97) [inline providers/google/tests/unit/google/cloud/hooks/test_gcs.py:1918]: Same here. See comment above (ensure both asserts are present).

File: providers/google/tests/unit/google/cloud/hooks/test_gcs.py [hunk 4]
```diff
@@ -1843,3 +1846,88 @@ def get_logs_string(call_args_list):
...
+        with pytest.raises(AirflowException, match="Insufficient disk space"):
+            hook.download(bucket_name="bucket", object_name="object", filename="/tmp/file")
+
+        # Verify blob.reload() was called to get the size
+        mock_blob.reload.assert_called_once()
+
+    @mock.patch(GCS_STRING.format("GoogleBaseHook.__init__"), new=mock_base_gcp_hook_default_project_id)
+    @mock.patch(GCS_STRING.format("GCSHook.get_conn"))
+    @mock.patch("shutil.disk_usage")
+    def test_download_works_if_enough_space(self, mock_disk_usage, mock_get_conn):
```
Comment (SameerMesiah97) [inline providers/google/tests/unit/google/cloud/hooks/test_gcs.py]: I think this test should handle boundary conditions i.e. `blob.size == free_space` and optionally 0. You could parameterize it by setting `mock_blob.size` to 1000, 2000 and 0.

**Review by SameerMesiah97 — COMMENTED** [review]
Looks good. I have left a few comments that you could address.

Please fix CI as welll.

**Comment by Srabasti** [comment]
Looks like trailing whitespaces at end of line are causing static checks to fail in multiple places @MonsterChenzhuo !

Can you run prek locally first and then commit from your branch?
Prek fixes any formatting errors, automatically.

File: providers/google/tests/unit/google/cloud/hooks/test_gcs.py (hunk 4 above)
Comment (MonsterChenzhuo) [reply providers/google/tests/unit/google/cloud/hooks/test_gcs.py]: done.

File: providers/google/src/airflow/providers/google/cloud/hooks/gcs.py (hunk 1 above)
Comment (MonsterChenzhuo) [reply providers/google/src/airflow/providers/google/cloud/hooks/gcs.py]: done.

File: providers/google/tests/unit/google/cloud/hooks/test_gcs.py (hunk 2 above)
Comment (MonsterChenzhuo) [reply providers/google/tests/unit/google/cloud/hooks/test_gcs.py:1877]: done.

File: providers/google/tests/unit/google/cloud/hooks/test_gcs.py (hunk 3 above)
Comment (MonsterChenzhuo) [reply providers/google/tests/unit/google/cloud/hooks/test_gcs.py:1918]: done.

**Comment by MonsterChenzhuo** [comment]
@SameerMesiah97 @Srabasti PTAL, thanks.

File: providers/google/tests/unit/google/cloud/hooks/test_gcs.py (hunk 2 above)
Comment (MonsterChenzhuo) [reply providers/google/tests/unit/google/cloud/hooks/test_gcs.py]: done.

File: providers/google/src/airflow/providers/google/cloud/hooks/gcs.py (hunk 1 above)
Comment (MonsterChenzhuo) [reply providers/google/src/airflow/providers/google/cloud/hooks/gcs.py]: done.

**Comment by MonsterChenzhuo** [comment]
@shahar1 PTAL, thanks.

**Comment by shahar1** [comment]
> @shahar1 PTAL, thanks.

It will take me a long while until I'll be able to review it properly due to low availability.
Feel free to work on other issues in the meanwhile.

File: providers/google/src/airflow/providers/google/cloud/hooks/gcs.py [hunk 5]
```diff
@@ -347,6 +344,16 @@ def download(
...
                 if filename:
+                    blob.reload(timeout=timeout)
+                    if blob.size is not None:
+                        directory = os.path.dirname(filename) or os.getcwd()
+                        free_space = shutil.disk_usage(directory).free
+                        if free_space < blob.size:
+                            raise AirflowException(
+                                f"Insufficient disk space to download file. "
+                                f"Required: {blob.size} bytes, Available: {free_space} bytes."
+                            )
```
Comment (shahar1) [inline providers/google/src/airflow/providers/google/cloud/hooks/gcs.py:355]: Could you please replace `AirflowException` with a native Python exception? (static checks should now fail because of that upon rebase/merge from `main`)

File: providers/google/src/airflow/providers/google/cloud/hooks/gcs.py [hunk 6]
```diff
@@ -347,6 +344,16 @@ def download(
                 blob = bucket.blob(blob_name=object_name, chunk_size=chunk_size)
 
                 if filename:
+                    blob.reload(timeout=timeout)
```
Comment (shahar1) [inline providers/google/src/airflow/providers/google/cloud/hooks/gcs.py:347]: 1. Can we extract the `blob.reload(...)` outside the retry loop?
2. Could you please add a flag (+docstring) for checking disk space (default: `False`), as it is an extra API call which might not be relevant when there are no storage limitations.

File: providers/google/src/airflow/providers/google/cloud/hooks/gcs.py [hunk 7]
```diff
@@ -347,6 +344,16 @@ def download(
...
+                    if blob.size is not None:
+                        directory = os.path.dirname(filename) or os.getcwd()
+                        free_space = shutil.disk_usage(directory).free
```
Comment (shahar1) [inline providers/google/src/airflow/providers/google/cloud/hooks/gcs.py:350]: It will fail with Python's `FileNotFoundError` if directory doesn't exist, rather than what GCS client raises.

**Review by shahar1 — CHANGES_REQUESTED** [review]
Apologies that it took too long - there are some comments to handle after merging from main.

**Comment by potiuk** [comment]
@MonsterChenzhuo — A reviewer (@shahar1) has requested changes on this PR, so I've removed the `ready for maintainer review` label — the next step is on your side. Could you address the review comments (push a fix, or reply in-thread explaining why the feedback doesn't apply)? Once addressed, re-request review from @shahar1 or re-mark the PR ready and it returns to the maintainer queue. Thank you.

_Note: This comment was drafted by an AI-assisted triage tool and may contain mistakes. Once you have addressed the points above, an Apache Airflow maintainer — a real person — will take the next look at your PR. We use this [two-stage triage process](https://github.com/apache/airflow/blob/main/contributing-docs/25_maintainer_pr_triage.md#why-the-first-pass-is-automated) so that our maintainers' limited time is spent where it matters most: the conversation with you._
